src-tauri/src/note.rs

```diff
@@ -13,26 +14,97 @@ impl Note {
-    pub fn list(&self) -> Vec<String> {
-        let base_path = self.config.get_base_path().join(".notes");
+    pub fn list(&self) -> Result<Vec<String>, NoteError> {
+        let base_path = self.config.get_base_path();
         match get_note_names(base_path) {
-          Ok(note_names) => return note_names,
+          Ok(note_names) => return Ok(note_names),
           Err(e) => error_handling(e.to_string())
         }
-        Vec::<String>::new()      
+        Ok(Vec::<String>::new())
     }
 
+    /// Reads the contents of the given note.
+    pub fn read(&self, name : &str) -> NoteResult<String> {
+      let readme = Path::new(&self.get_note_path(name)).join("README.md");
+      Ok(fs::read_to_string(readme)?)
+    }
+
+    /// Creates a new note and return it's name.
+    pub fn create(&self) -> NoteResult<String> {
+      let mut name = String::from("Untitled");
+      let mut path = self.get_note_path(&name);
+      let mut n = 0;
+      while path.as_path().exists() {
+        n += 1;
+        name = format!("Untitled {}", n);
+        path = self.get_note_path(&name);
+      }
+
+      fs::create_dir(path.clone())?;
+      let readme = Path::new(&path).join("README.md");
+      fs::write(readme, format!("# {}\n", name))?;
+      let tags = Path::new(&path).join("tags.txt");
+      fs::write(tags, "")?;
+
+      Ok(name)
+    }
+
+    /// Renames a note.
+    pub fn rename(&self, old_name: &str, new_name: &str) -> NoteResult<()> {
+      let old_path = self.get_note_path(old_name);
+      let new_path = self.get_note_path(new_name);
+      Ok(fs::rename(old_path, new_path)?)
+    }
+
+    /// Writes the content of the given notes.
+    pub fn write(&self, name: &str, content: &str) -> NoteResult<()> {
+      let readme = Path::new(&self.get_note_path(name)).join("README.md");
+      fs::write(readme, content.as_bytes())?;
+      Ok(())
+    }
+
+    /// Removes the given note.
+    pub fn remove(&self, name: &str) -> NoteResult<()> {
+      let path = self.get_note_path(name);
+      Ok(fs::remove_dir_all(path)?)
+    }
+
+    /// Reads the tags associated with the given note.
+    pub fn read_tags(&self, name: &str) -> NoteResult<Vec<String>> {
+      let tags = Path::new(&self.get_note_path(name)).join("tags.txt");
+      let content = fs::read_to_string(tags)?;
+      let tags: Vec<String> = content
+        .lines()
+        .filter(|s| !s.is_empty())
+        .map(String::from)
+        .collect();
+      Ok (tags)
+    }
+
+    /// Replaces the tags of tags of the given note.
+    pub fn write_tags(&self, name: &str, tags: &Vec<String>) -> NoteResult<()> {
+      let content = tags.join("\n");
+      let tags = Path::new(&self.get_note_path(name)).join("tags.txt");
+      fs::write(tags, content.as_bytes())?;
+      Ok(())
+    }
+
+    fn get_note_path(&self, name: &str) -> PathBuf {
+      let mut path = self.config.get_base_path();
+      path.push(name);
```

Comment:
I don't know if the described behaviour for a name with a starting `/` or `\` is what you want to accept here :https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.push

Reply:
Good point. I'll fix this.

Reply:
The `check_name` function was added to address this.